### PR TITLE
feat: Add initial OCaml support

### PR DESCRIPTION
**Pull request by wwared**
This PR introduces OCaml support by extracting the [lambda IR](https://github.com/ocaml/ocaml/blob/1ecb8f7327dd0837d899bef60cd9e9a87219ffb6/lambda/lambda.ml#L292) from the ocaml compiler, parsing it into a rust enum, then transforming it into an equivalent lurk program. Note that a lot of the behavior will change and this is still experimental and incomplete.

This adds two new meta commands: `!(load-ocaml)` which functions like `!(load)` and `!(load-ocaml-expr)` which functions like `!(load-expr)`. The former will run the program and print its result, while the latter returns the transformed lurk program.

You can do things like `!(load-ocaml "file.ml") !(debug) !(prove)`, `!(debug !(load-ocaml-expr "file.ml"))` or `!(defq program !(load-ocaml-expr "file.ml"))`. The file must end in a `.ml` extension, and currently we only support compiling a single file.

Due to the way individual `.ml` files are compiled, for now you will want a program like `let result = 123` rather than `123`. Otherwise the `123` is evaluated inside a `begin` and its result is not used.

The pipeline works as follows:
* Create a temporary directory using [tempfile](https://docs.rs/tempfile/latest/tempfile/) to hold build artifacts. `ocamlc` creates `.cmi` and `.cmo` files that are not relevant to us.
* Call `ocamlc -dlambda -dno-unique-ids -warn-error +a -c <file>`. The code must compile with no warnings. `-dlambda` makes the compiler output the lambda IR to stderr. `-dno-unique-ids` removes the unique suffixes added to identifiers by the compiler.
* Parse the lambda IR into a `LambdaSyntax` struct. Forms that are not supported get parsed as `FallbackPrimitive`s or `FallbackLiteral`s.
* Generate a lurk program of the form `(let (<helper functions>) <ocaml expr>)` using `transform_lambda`.

The process for supporting a new primitive is roughly:
* Add an entry to `LambdaSyntax`
* Modify `ocaml::parser::syntax::parse_primitive_sexp` to correctly parse the sexp into the new type (optionally add a parser test)
* Handle the case in `ocaml::compile::transform_lambda`
* Add a test case to `lurk::tests::eval_ocaml`

Currently we support:
* Basic data types: ints, chars, strings, records (no field access yet), functions
* `let` and `let rec`. `seq` as `begin`. function calls
* Primitives with a direct lurk version: `=`, `>`, etc (though ocaml supports comparing chars and strings with these)

Records, arrays, lists, tuples and etc will initially be represented by simple lists and accessed with `nth`. A record `[0: 123 "abc"]` becomes `(list 0 123 "abc")`, so a record is a list of its tag followed by its members.

It's possible the `-dno-unique-ids` option could lead to issues in certain programs, but for now we're using it by default. We might need to revisit this and use a different solution if we find a problematic program with this flag.

OCaml tests can be added inline or as files to `src/lurk/tests/ml/` and a corresponding `test_file!` call in `tests/eval_ocaml.rs`.

Known issues/limitations:
* OCaml `true`/`false` are represented sometimes by `1`/`0` ints, which can lead to problems if used as conditions with `if`
* Floats are not supported and will add explicit `(fail)`s in the lurk program
* Calling stdlib functions does not work
* Integers are represented as `u64`s and negative integers are not supported yet
* `match` statements are not supported

### CODE REVIEW

File: src/lurk/cli/meta.rs [hunk 1]
```diff
@@ -1441,6 +1444,40 @@ impl<C1: Chipset<F>, C2: Chipset<F>> MetaCmd<F, C1, C2> {
             Ok(*repl.zstore.t())
         },
     };
+
+    const OCAML_LOAD: Self = Self {
+        name: "ocaml-load",
+        summary: "(Experimental) Load OCaml expressions from a file.",
+        format: "!(ocaml-load <string>)",
+        info: &[],
+        example: &["!(load \"my_file.ml\")"],
```
Comment (arthurpaulino) [inline src/lurk/cli/meta.rs:1453]: Nit: info comes before format

Fix: use `ocaml-load` in the example

File: src/ocaml/compile.rs [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    let file_name = orig_path
+        .file_name()
+        .ok_or(anyhow!("Invalid file name: {}", orig_path))?;
+    compile_single_file_contents(file_contents.as_str(), file_name)
+}
+
+/// Compiles a single file with `ocamlc` and returns the resulting lambda IR.
+///
+/// This writes the data to a temporary file in a temporary directory, runs
+/// `ocamlc -dlambda -o /dev/null <file>` and captures the stderr.
```
Comment (arthurpaulino) [inline src/ocaml/compile.rs:30]: Is the `/dev/null` part up to date? I think it needs an update

File: src/ocaml/parser/syntax.rs [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+        )),
+        multispace0,
+    )(from)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn set_pos(l: &LambdaSyntax, p: Pos) -> LambdaSyntax {
```
Comment (arthurpaulino) [inline src/ocaml/parser/syntax.rs:261]: It's cleaner if you implement equality by hand, ignoring the `Pos`, similar to how it was done for `Syntax`. You can also implement `Hash` following the same idea (it was also done for `Syntax`).

File: src/ocaml/parser/syntax.rs [hunk 4]
```diff
@@ -0,0 +1,482 @@
...
+            fn $test_func() {
+                let placeholder_span = Span::new("");
+                let p = Pos::from_upto(placeholder_span, placeholder_span);
+                let expected = ($expected_cloj)(p);
+                test_parser($input_str, Some(expected))
+            }
+        };
+    }
+
+    test!(test_mastermind, include_str!("tests/mastermind.ir"));
```
Comment (arthurpaulino) [inline src/ocaml/parser/syntax.rs:339]: It's probably worth mentioning the version of the compiler used to generate this IR

**Review by arthurpaulino — COMMENTED** [review]
Just minor comments, nothing critical.
Thanks again!

File: src/ocaml/parser/syntax.rs (hunk 3 above)
Comment (wwared) [reply src/ocaml/parser/syntax.rs:261]: D'oh, this is so much nicer! I don't even need this `set_pos` stuff at all

File: src/ocaml/parser/syntax.rs (hunk 4 above)
Comment (wwared) [reply src/ocaml/parser/syntax.rs:289]: I've also added the original ocaml source file in the test folder for reference, though right now it's not being used as a test yet

**Review by arthurpaulino — APPROVED** [review]
🚀

**Pull request merged**